Feature/pagination

spacelessvoid opened the pull request:
Добавил компоненты пагинации (стили и логику рендеринга).
Для мобильной верстки сделал допущение, что в блоке пагинации отображаются минимум 3 страницы (из макета точно не понятно, потому что видно только первую и последнюю).

spacelessvoid committed:
feat: добавил верстку и стиль компонентов пагинации

spacelessvoid committed:
добавил конфигурацию форматирования в .prettierrc

spacelessvoid committed:
добавил в компонент PaginationItem условие: не отображрать ссылку, если он выбран

spacelessvoid committed:
написал логику рендеринга элементов пагинации + внес необходимые правки в компоненты пагинации

spacelessvoid committed:
Squashed commit of the following:

commit dc5800f4cdab4b7068b728ce15cae1ff1d5dd6dc
Merge: 69a6497 9968c95
Author: AlMkin <138128998+AlMkin@users.noreply.github.com>
Date:   Tue Mar 26 19:22:57 2024 +0300

    Merge pull request #11 from Trudovyereservy/feature/eslint-import-sort

    Feature/eslint import sort

commit 9968c95a085ea6e5ae97179ad84a5da988322b81
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Mon Mar 25 19:48:57 2024 +0700

    feat: добавил правило в линтер ограничивающее количество пустых строк до максимум одной

commit 7efd298f956bb283ccdcff493597ab910156fb34
Merge: ab25abe 69a6497
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Mon Mar 25 19:08:50 2024 +0700

    Merge branch 'develop' of github.com:Trudovyereservy/frontend_part into feature/eslint-import-sort

commit ab25abef08e032e791c443f5ed56a976f6dc8a41
Merge: 92f993d 62bce03
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 14:18:59 2024 +0700

    Merge branch 'feature/eslint-import-sort' of github.com:Trudovyereservy/frontend_part into feature/eslint-import-sort

commit 92f993d65e05b87cad99d6bfc4d30440b0ab2162
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 14:13:33 2024 +0700

    feat: добавил правило пустой строки после директив и исправил везде yarn lint --fix

commit 62bce034bcde734c48eb32226d74219434fafd51
Merge: aa8b781 359e06a
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 13:16:21 2024 +0700

    Merge branch 'feature/eslint-import-sort' of github.com:Trudovyereservy/frontend_part into feature/eslint-import-sort

commit aa8b781f036187f45c85dc674460784ee3eab89a
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 13:16:09 2024 +0700

    fix: поправил название constant в импорте, добавил правило пустой строки после директив

commit 69a6497d07b4793560cb4b191aa99bbe9a2310da
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Sat Mar 23 01:35:51 2024 +0300

    Поправил имя constants, поправил по мелочи файлы

commit 359e06a0baf2441390a1627095fdb66f7fd434e4
Merge: 6ba9071 14b64e3
Author: Glushenkov Konstantin <glukosalex@gmail.com>
Date:   Thu Mar 21 22:05:02 2024 +0700

    Merge branch 'develop' into feature/eslint-import-sort

commit 14b64e3888846e4ca67ea3c044f2db2a9003d592
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Thu Mar 21 00:09:24 2024 +0300

    Сделал один файл Constans, удалил в форме картинки по 7 МБ, скорректировал компонент Input, в константах поправил уникальность id

commit 6ba90717a5982463bd786474dce1c228b066de5d
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 23:07:03 2024 +0700

    fix: fix path to images

commit e9c3a247af466cf70f1b6aaa6d0e920a1cfb2343
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 23:02:57 2024 +0700

    style: fix parent paths to alias '@'

commit d70a6801eefa56f4386efacc6f952c08d997ed0b
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 23:02:04 2024 +0700

    style: add eslint rule import/newline-after-import

commit fba2499ac5f1e5d85bce8af328f4d8ad8e8ca0b7
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 22:33:12 2024 +0700

    style: auto sort all imports

commit c56c179b63378a53ba1fce00b7b4064aa7343af9
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 22:32:31 2024 +0700

    style: config eslint and prettier

commit f969aecdaff18e5ac8ad0b770885a0bfc63579c9
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 22:31:08 2024 +0700

    build: install simple-import-sort plugin

spacelessvoid committed:
установил новые зависимости после мерджа из ветки develop

spacelessvoid committed:
style: добавил компоненту пагинации стили позиционирования

также обновил стили компонентов на связанных страницах (BlockCards\CardsList, CoachList)

spacelessvoid committed:
fix: исправил constResizeWindow.ts

добавил брейкпойнты в соответствии с макетом и _variables.scss; количество карточек для брейкпойнтов указал в соответствии с макетом версии 2

spacelessvoid committed:
fix: исправил useCardCount.ts

количество карточек рендерится как в макете: 8 шт. для мобилок, 9 шт. в остальных случаях и по умолчанию

spacelessvoid committed:
адаптировал компонент пагинации под разрешение телефона

AlMkin commented on src/components/Pagination/Pagination.module.scss:
```diff
@@ -0,0 +1,110 @@
+}
+
+.pagination__chevron {
+  background-image: $chevron;
```
думаю можно вынести в mixin
@mixin chevron-background($image) {
  background-image: $image;
  background-position: center;
  background-repeat: no-repeat;
}

AlMkin commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+    const siblings = (paginationItemsToDisplay - 3) / 2;
+
+    // getting an array with page numbers to display
+    function range(start: number, end: number) {
```
range можно вынести в utils и немного разгрузить renderPaginationItems

AlMkin commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+  const totalPages = Math.ceil(totalCards / cardsPerPage);
+  const lastPage = totalPages;
+
+  function renderPaginationItems() {
```
вызывать функции в return (в render функции) не желательно, в доке по react кажется есть это рекомендация

давай это завернем в useMemo

AlMkin commented on src/components/Pagination/PaginationNext.tsx:
```diff
@@ -0,0 +1,26 @@
+  lastPage: number;
+};
+
+export const PaginationNext = ({ href, currentPage, lastPage }: PaginationNextProps) => {
```
у тебя Next и Prev одинаковые почти, сделай один PaginationArrow и пропсом указывай направление
типа diraction="prev" | "next"

AlMkin reviewed (changes requested):
👍🏽

GluKosAlex commented on src/components/Pagination/Pagination.module.scss:
```diff
@@ -0,0 +1,110 @@
+@import '../../styles/_mixins.scss';
+
+$chevron: url('../../../public/chevron\ .\ arrows.svg');
+$chevron-inverse: url('../../../public/chevron\ .\ arrows.grey.svg');
```
Имена файлов с пробелами и точками выглядят жутковато. В путях приходится экранировать и читается это не очень. Я бы посоветовал унифицировать именование файлов или в snake_case, или в kebab-case, или аля по БЭМ (chevron-arrows.svg, chevron-arrows__active.svg)

GluKosAlex commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+    if (totalPages > paginationItemsToDisplay && currentPage + siblings * 2 >= totalPages) {
+      return (
+        <>
+          <PaginationItem href={`?limit=${cardsPerPage}&page=${1}`} pageNumber={1} />
```
Возможно, PaginationItem придётся сделать не ссылкой, а кнопкой, тк будет вызываться функция с фетчем, которая будет принимать limit и page, и передавать их в запрос в качестве query параметров. Или ссылке придутся указывать onClick и пресекать дефолтное поведение.

spacelessvoid replied:
Пока что не стал переделывать в кнопку, потому что `Link` использовал подразумевая, что элементы будут листаться без полного перерендера страницы. Но, возможно, так не получится сделать - я себе этот момент отметил, чтобы в будущем вернуться к этому и, если что, переделать в кнопку

GluKosAlex replied:
Link - это хорошо, но он для роутинга по приложению. У нас по запросу с данными урлами + квери параметрами на апи бэка приходит объект с пагинированным массивом, который надо будет отрисовать. Я тебя просто хочу предостеречь от лишней работы. Ты сейчас повыносишь всякие там методы по генерации урлов в отдельные файлы, а они по сути не понадобятся, так как тебе всего лишь придётся вызвать сервер акшн по онКлик и передать туда несколько параметров, включающих в себя лимит и номер страницы. Просто напоминаю - у нас пагинация на стороне бэка.

spacelessvoid replied:
В итоге методы по генерации урлов я пока не наделал =) Думаю, они и не нужны будут.
Спасибо за предостережения, Кость (без иронии).
Меня твой коммент снова навел на мысли о том, как данные должны приходить и я нарисовал схемку, чтобы не забыть.
По сути, нажимая на ссылку, мы ничего никуда не отправляем - это не нужно. У нас загоняются параметры в адресную строку, а страница на основе этого обращается на сервер за контентом.
Насколько я понимаю, это вполне себе в рамках парадигмы Некста.
Только у нас страницы 'use client', а мне кажется что должны быть серверные.

Схема (рисовал для себя, поэтому не пытался сильно понятно сделать):
![2024-04-29 - TR-data-fetching](https://github.com/Trudovyereservy/frontend_part/assets/55495/2103bd4d-4cc4-4f90-969f-9c3579800d23)

AlMkin commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+    if (totalPages <= paginationItemsToDisplay) {
+      return (
+        <>
+          {range(1, totalPages).map((page) => (
```
давай ещё вынесим все PaginationItem в PaginationList, чтобы разгрузить компонент
```
          {range(1, totalPages).map((page) => (
            <PaginationItem
              href={`?limit=${cardsPerPage}&page=${page}`}
              pageNumber={page}
              currentPage={currentPage}
              key={page}
            />
          ))}
```

AlMkin commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+    if (totalPages > paginationItemsToDisplay && currentPage + siblings * 2 >= totalPages) {
+      return (
+        <>
+          <PaginationItem href={`?limit=${cardsPerPage}&page=${1}`} pageNumber={1} />
```
можно вынести в utils `?limit=${cardsPerPage}&page=${1}`

spacelessvoid replied:
Не стал выносить в `utils`, потому что строки немного разные в зависимости от условий. Повторяющаяся строка была внутри PaginationItem и ушла в PaginationList в единственном экземпляре

AlMkin replied:
тут я имел ввиду просто

export function createPaginationUrl(cardsPerPage, page) {
  return `?limit=${cardsPerPage}&page=${page}`;
}

кажется он везде одинаковый

spacelessvoid replied:
Не совсем так: он был одинаковый в нескольких местах, но теперь именно в таком виде строчка только в компоненте PaginationList, в других немного разные вариации. Как функция, возвращающая строку, будет работать, в принципе, но, опять же, по-моему излишняя абстракция.
![image](https://github.com/Trudovyereservy/frontend_part/assets/55495/8976221e-3589-42c0-80c6-b83626102597)

AlMkin replied:
я говорю про то что у тебя строка одинаковая, ты туда просто передаешь limit и page

сейчас может показаться как лишняя абстракция, но это не так
потом придут к тебе и попросят добавить ещё один какой-то параметр, а потом ещё придут и добавить надо будет сортировку какую-нить это ещё пара параметров, а потом переименовать может какую нить параметр захочется, чтобы с другим конектилось и тебе уже не в одном месте все это менять придётся

в общем, по моему опыту это надо выносить

spacelessvoid replied:
Я в итоге вообще зарефакторил эту часть =) Осталось `href={'?page=${1}'}`
Решил, что в ссылке не нужен параметр `limit`, мы его можем в компоненте расчитывать и передавать функции, которая делает запрос на сервер.

AlMkin commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    // only the left dots are visible
+    if (totalPages > paginationItemsToDisplay && currentPage + siblings * 2 >= totalPages) {
```
такие тяжелые условия тоже лучше вынести в utils

spacelessvoid replied:
Вопрос по этому и предыдущему комментарию (к 58-й строке): такая дополнительная абстракция не усложнит чтение и понимание файла? 
Может быть тогда renderPaginationItems() целиком вынести в отдельный хук?

AlMkin replied:
нет, только улучшит если красиво назовешь, так как у тебя тут аж 5 операторов, больше трёх уже можно выносить
по 58, не критично, но я бы вынес, принцип DRY ещё не отменяли) проще будет если захочется, что-то изменить/добавить, в одном месте сделаешь, а не в нескольких

по поводу renderPaginationItems: я наверху оставил коммент, что можно вынести PaginationItem в List, в хуки мы все таки выносим больше логику
по итогу у тебя в Pagination останется только условия отрисовки ui, а логику и повторяющиеся вещи выносим

spacelessvoid replied:
В итоге тоже не вынес в `utils`, потому что это потребовало бы импортировать переменные, которые высчитваются в компоненте - посчитал это нецелесообразным.
Но вынес условия в константы внутри компонента для чуть лучшей читаемости

AlMkin commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,132 @@
+
+export const Pagination = ({ totalCards, currentPage }: PaginationItemProps) => {
+  const width: number = useWindowSize();
+  const cardsPerPage: number = useCardCount(width);
```
вот это все можно вынести в custom hook и вытащить из  него все, что нам нужно, старайся отделять бизнес логику от ui, и в идеале в компоненте не должно быть голых чисел `BREAKPOINT_TABLET ? 5 : 3` 5, 3 лучше выносить в константы

```
  const width: number = useWindowSize();
  const cardsPerPage: number = useCardCount(width);

  const paginationItemsToDisplay = width > BREAKPOINT_TABLET ? 5 : 3;

  const totalPages = Math.ceil(totalCards / cardsPerPage);
  const lastPage = totalPages;
```

AlMkin commented on src/components/Pagination/PaginationItem.tsx:
```diff
@@ -0,0 +1,34 @@
+};
+
+export const PaginationItem = ({ href, pageNumber, currentPage }: PaginationItemProps) => {
+  const isSelected = pageNumber === currentPage ? true : false;
```
тут необязательно явно возвращать true и false, pageNumber === currentPage - достаточно

spacelessvoid committed:
Squashed commit of the following:

commit d4ae7258aa50750253cd4b247303de252ab55ce5
Merge: fe7cf43 8c26dcc
Author: AlMkin <138128998+AlMkin@users.noreply.github.com>
Date:   Sat Apr 13 15:55:33 2024 +0300

    Merge pull request #13 from Trudovyereservy/feature/contacts-components

    Feature/contacts components

commit 8c26dcc55de5187487608ebe0c6e3da597b0c5f5
Merge: 7c7dbf8 fe7cf43
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Thu Apr 11 04:26:28 2024 +0700

    Merge branch 'develop' into feature/contacts-components

commit 7c7dbf88751fd00d1c413fd093b9273cba8c6236
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Thu Apr 11 04:25:13 2024 +0700

    fix: вынес переменные в стилях

commit 49e37a9726dae7b9b7d820a7128a16d92f48b9f0
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Mon Apr 8 04:36:57 2024 +0700

    fix: исправил нейминг стилей

commit 5c7ff20f4bccd1b137f77af5710e1aa51003c0af
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Mon Apr 8 04:34:18 2024 +0700

    fix: исправил ошибки линтера

commit fe7cf4329bf6c23db5744a62472f19b6dbd1ca06
Author: Aleksandr Mukhin <aamukhin@alfabank.ru>
Date:   Sun Apr 7 11:35:11 2024 +0800

    feat(add-eslint-rule): add eslint rules

commit 1989b30ae2a2a2d8dc1e283594c2297899eae6c4
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 22:50:38 2024 +0700

    feat: добавил переключение карт по клику по заголовку

commit edc8e006db87f36fd60e68db4c6a5fe4a18d49f6
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 22:25:15 2024 +0700

    feat: добавил карточки контактов

commit d163c3ec2eea3bc99b7b1e722c4abeeddde9cabe
Merge: e525cb9 dc5800f
Author: Aleksandr Mukhin <aamukhin@alfabank.ru>
Date:   Tue Apr 2 20:16:29 2024 +0800

    Merge branch 'develop' of https://github.com/Trudovyereservy/frontend_part into develop

commit 204aad99c0a8be5bccd3f7d0eb2f2f05952c8845
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 06:29:01 2024 +0700

    feat: добавил лист адресов

commit d3ddabde7d72d98abd69e68f8420b967786ffd06
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 03:03:07 2024 +0700

    addresses initial commit

commit 457cba7bda18f44508f5f8c7940f7d0760c554ed
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 02:57:11 2024 +0700

    feat: добавил карту

commit 40171e15c134f82eaadab71b6fc1c620f66f4461
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Fri Mar 29 04:14:28 2024 +0700

    initial commit

commit dc5800f4cdab4b7068b728ce15cae1ff1d5dd6dc
Merge: 69a6497 9968c95
Author: AlMkin <138128998+AlMkin@users.noreply.github.com>
Date:   Tue Mar 26 19:22:57 2024 +0300

    Merge pull request #11 from Trudovyereservy/feature/eslint-import-sort

    Feature/eslint import sort

commit 9968c95a085ea6e5ae97179ad84a5da988322b81
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Mon Mar 25 19:48:57 2024 +0700

    feat: добавил правило в линтер ограничивающее количество пустых строк до максимум одной

commit 7efd298f956bb283ccdcff493597ab910156fb34
Merge: ab25abe 69a6497
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Mon Mar 25 19:08:50 2024 +0700

    Merge branch 'develop' of github.com:Trudovyereservy/frontend_part into feature/eslint-import-sort

commit ab25abef08e032e791c443f5ed56a976f6dc8a41
Merge: 92f993d 62bce03
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 14:18:59 2024 +0700

    Merge branch 'feature/eslint-import-sort' of github.com:Trudovyereservy/frontend_part into feature/eslint-import-sort

commit 92f993d65e05b87cad99d6bfc4d30440b0ab2162
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 14:13:33 2024 +0700

    feat: добавил правило пустой строки после директив и исправил везде yarn lint --fix

commit 62bce034bcde734c48eb32226d74219434fafd51
Merge: aa8b781 359e06a
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 13:16:21 2024 +0700

    Merge branch 'feature/eslint-import-sort' of github.com:Trudovyereservy/frontend_part into feature/eslint-import-sort

commit aa8b781f036187f45c85dc674460784ee3eab89a
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 13:16:09 2024 +0700

    fix: поправил название constant в импорте, добавил правило пустой строки после директив

commit 69a6497d07b4793560cb4b191aa99bbe9a2310da
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Sat Mar 23 01:35:51 2024 +0300

    Поправил имя constants, поправил по мелочи файлы

commit 359e06a0baf2441390a1627095fdb66f7fd434e4
Merge: 6ba9071 14b64e3
Author: Glushenkov Konstantin <glukosalex@gmail.com>
Date:   Thu Mar 21 22:05:02 2024 +0700

    Merge branch 'develop' into feature/eslint-import-sort

commit 14b64e3888846e4ca67ea3c044f2db2a9003d592
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Thu Mar 21 00:09:24 2024 +0300

    Сделал один файл Constans, удалил в форме картинки по 7 МБ, скорректировал компонент Input, в константах поправил уникальность id

commit 6ba90717a5982463bd786474dce1c228b066de5d
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 23:07:03 2024 +0700

    fix: fix path to images

commit e9c3a247af466cf70f1b6aaa6d0e920a1cfb2343
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 23:02:57 2024 +0700

    style: fix parent paths to alias '@'

commit d70a6801eefa56f4386efacc6f952c08d997ed0b
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 23:02:04 2024 +0700

    style: add eslint rule import/newline-after-import

commit fba2499ac5f1e5d85bce8af328f4d8ad8e8ca0b7
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 22:33:12 2024 +0700

    style: auto sort all imports

commit c56c179b63378a53ba1fce00b7b4064aa7343af9
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 22:32:31 2024 +0700

    style: config eslint and prettier

commit f969aecdaff18e5ac8ad0b770885a0bfc63579c9
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 22:31:08 2024 +0700

    build: install simple-import-sort plugin

spacelessvoid committed:
добавил миксин в Pagination.module.scss + переименовал файлы изображений с шевроном

spacelessvoid committed:
объединил PaginationNext & PaginationPrevious в один компонент PaginationArrow

spacelessvoid committed:
вынес range() в отдельный файл pagination.ts

spacelessvoid committed:
вынес часть логики из компонента Pagination в новый хук usePagination()

spacelessvoid committed:
вынес в отдельный компонент PaginationList функцию отрисовки списка элементов

spacelessvoid committed:
вынес условия рендеринга элементов пагинации в константы

spacelessvoid committed:
добавил мемоизацию для элементов пагинации

spacelessvoid committed:
добавил компонент пагинации на релевантные страницы - пока без функциональности

AlMkin commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,119 @@
+  // number of items to display before and after the current page
+  const siblings = (paginationItemsToDisplay - 3) / 2;
+
+  const ellipsisNotVisible = totalPages <= paginationItemsToDisplay;
```
разбивай сложные условия на более простые логические переменные, не больше трёх условий + нейминг

`const isPaginationNecessary = totalPages > paginationItemsToDisplay;
const isCurrentPageNearEnd = currentPage + siblingCount * 2 >= totalPages;
const isCurrentPageNearStart = currentPage <= siblingCount + 2;
const isCurrentPageAwayFromBothEnds = currentPage > siblingCount * 2 + 1;

const ellipsisNotVisible = !isPaginationNecessary;
const ellipsisLeftVisible = isPaginationNecessary && isCurrentPageNearEnd;
const ellipsisRightVisible = isPaginationNecessary && isCurrentPageNearStart;
const ellipsisLeftAndRightVisible = isPaginationNecessary && isCurrentPageAwayFromBothEnds;`

AlMkin replied:
https://refactoring.com в книге Fowler-a можно найти полезные техники

AlMkin commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -0,0 +1,119 @@
+    usePagination(totalCards);
+
+  // number of items to display before and after the current page
+  const siblings = (paginationItemsToDisplay - 3) / 2;
```
const ANY_MAGIC_NUMBER = 3; // Должно пояснять что за '3'

const siblings = (paginationItemsToDisplay - ANY_MAGIC_NUMBER) / 2;

и тут Math.floor не надо?

AlMkin commented on src/components/Pagination/PaginationArrow.tsx:
```diff
@@ -0,0 +1,47 @@
+    direction === 'next'
+      ? styles.pagination__chevron_disabled
+      : styles.pagination__chevronInverse_disabled;
+
```
вот это все:
`  const isNextDisabled = currentPage === lastPage;
  const isPreviousDisabled = currentPage === 1;
  const isDisabled =
    (isNextDisabled && direction === 'next') || (isPreviousDisabled && direction === 'previous');

  const chevronStyle =
    direction === 'next' ? styles.pagination__chevron : styles.pagination__chevronInverse;
  const chevronDisabledStyle =
    direction === 'next'
      ? styles.pagination__chevron_disabled
      : styles.pagination__chevronInverse_disabled;`

можно сократить тут до:

const isDisabled = getIsDisabled(direction, currentPage, lastPage);
const chevronStyles = getChevronStyles(direction, isDisabled);

spacelessvoid replied:
Ты имеешь ввиду, что `getIsDisabled` и `getChevronStyles` нужно сделать функциями, которые возвращают соответствующие условия?

AlMkin replied:
ага, сразу полегче будет компонент

spacelessvoid committed:
разбил условия отрисовки пагинации на более простые

spacelessvoid committed:
вынес условия навешивания стилей на PaginationArrow в отдельные функции getIsChevronDisabled и getChevronStyles

spacelessvoid committed:
refactor: убрал параметр cardsPerPage из ссылок в элементах пагинации

данный параметр расчитывается на странице и передается в fetch запрос к серверу, его не обязательно делать эксплицитным для пользователя

spacelessvoid committed:
Squashed commit of the following:

commit 0da20781a3c21bff3b78b57ed68b535a2c3914a5
Merge: d4ae725 fd476fb
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Mon Apr 29 08:32:00 2024 +0300

    Merge pull request #15 from Trudovyereservy/feature/SwiperPost

    Feature/swiper post

commit fd476fbef03cfb819b84ef158289a996c6edfe3e
Merge: ae9b1b5 d4ae725
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Tue Apr 16 00:57:10 2024 +0300

    На проверку

commit ae9b1b5fd3599664d982c07fd26e7a6189bb333e
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Tue Apr 16 00:51:38 2024 +0300

    На проверку

commit d4ae7258aa50750253cd4b247303de252ab55ce5
Merge: fe7cf43 8c26dcc
Author: AlMkin <138128998+AlMkin@users.noreply.github.com>
Date:   Sat Apr 13 15:55:33 2024 +0300

    Merge pull request #13 from Trudovyereservy/feature/contacts-components

    Feature/contacts components

commit 8c26dcc55de5187487608ebe0c6e3da597b0c5f5
Merge: 7c7dbf8 fe7cf43
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Thu Apr 11 04:26:28 2024 +0700

    Merge branch 'develop' into feature/contacts-components

commit 7c7dbf88751fd00d1c413fd093b9273cba8c6236
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Thu Apr 11 04:25:13 2024 +0700

    fix: вынес переменные в стилях

commit 43cfcfc899ca8c1ab790aab144e20e28f4104564
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Wed Apr 10 23:13:35 2024 +0300

    Поправил замечания.Еще нужно доделать анимацию перехода

commit 8486bc00a6b5621fd8233255c805d59b95021ed1
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Mon Apr 8 00:58:41 2024 +0300

    Поправил замечания. сделаны стили. на проверку

commit 49e37a9726dae7b9b7d820a7128a16d92f48b9f0
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Mon Apr 8 04:36:57 2024 +0700

    fix: исправил нейминг стилей

commit 5c7ff20f4bccd1b137f77af5710e1aa51003c0af
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Mon Apr 8 04:34:18 2024 +0700

    fix: исправил ошибки линтера

commit beed34eec369b1d7c8f666ccba5a057e41918808
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Sun Apr 7 19:44:07 2024 +0300

    Поправил замечания. стили засада

commit fe7cf4329bf6c23db5744a62472f19b6dbd1ca06
Author: Aleksandr Mukhin <aamukhin@alfabank.ru>
Date:   Sun Apr 7 11:35:11 2024 +0800

    feat(add-eslint-rule): add eslint rules

commit 16b7d8fc25a65b9b4c40161a11caa7b2bbb8feb7
Merge: 973edfd dc5800f
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Wed Apr 3 23:59:42 2024 +0300

    Поправил замечания. Осталось разобраться с брейкпоинтами

commit 1989b30ae2a2a2d8dc1e283594c2297899eae6c4
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 22:50:38 2024 +0700

    feat: добавил переключение карт по клику по заголовку

commit edc8e006db87f36fd60e68db4c6a5fe4a18d49f6
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 22:25:15 2024 +0700

    feat: добавил карточки контактов

commit d163c3ec2eea3bc99b7b1e722c4abeeddde9cabe
Merge: e525cb9 dc5800f
Author: Aleksandr Mukhin <aamukhin@alfabank.ru>
Date:   Tue Apr 2 20:16:29 2024 +0800

    Merge branch 'develop' of https://github.com/Trudovyereservy/frontend_part into develop

commit 204aad99c0a8be5bccd3f7d0eb2f2f05952c8845
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 06:29:01 2024 +0700

    feat: добавил лист адресов

commit d3ddabde7d72d98abd69e68f8420b967786ffd06
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 03:03:07 2024 +0700

    addresses initial commit

commit 457cba7bda18f44508f5f8c7940f7d0760c554ed
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Tue Apr 2 02:57:11 2024 +0700

    feat: добавил карту

commit 973edfd3e82fcd776dfd63190c78d7fe86916cfa
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Fri Mar 29 01:00:08 2024 +0300

    Исправл код по замеаниям код-ревью. На проверку.

commit 40171e15c134f82eaadab71b6fc1c620f66f4461
Author: Daniil Isaev <shandydenn@yandex.ru>
Date:   Fri Mar 29 04:14:28 2024 +0700

    initial commit

commit dc5800f4cdab4b7068b728ce15cae1ff1d5dd6dc
Merge: 69a6497 9968c95
Author: AlMkin <138128998+AlMkin@users.noreply.github.com>
Date:   Tue Mar 26 19:22:57 2024 +0300

    Merge pull request #11 from Trudovyereservy/feature/eslint-import-sort

    Feature/eslint import sort

commit 9968c95a085ea6e5ae97179ad84a5da988322b81
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Mon Mar 25 19:48:57 2024 +0700

    feat: добавил правило в линтер ограничивающее количество пустых строк до максимум одной

commit 7efd298f956bb283ccdcff493597ab910156fb34
Merge: ab25abe 69a6497
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Mon Mar 25 19:08:50 2024 +0700

    Merge branch 'develop' of github.com:Trudovyereservy/frontend_part into feature/eslint-import-sort

commit 966dcbff56afa322dd53e90b8f12e95cb2f2e339
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Sat Mar 23 23:23:50 2024 +0300

    Поправил по замечаниям. На проверку повторную

commit ab25abef08e032e791c443f5ed56a976f6dc8a41
Merge: 92f993d 62bce03
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 14:18:59 2024 +0700

    Merge branch 'feature/eslint-import-sort' of github.com:Trudovyereservy/frontend_part into feature/eslint-import-sort

commit 92f993d65e05b87cad99d6bfc4d30440b0ab2162
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 14:13:33 2024 +0700

    feat: добавил правило пустой строки после директив и исправил везде yarn lint --fix

commit 62bce034bcde734c48eb32226d74219434fafd51
Merge: aa8b781 359e06a
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 13:16:21 2024 +0700

    Merge branch 'feature/eslint-import-sort' of github.com:Trudovyereservy/frontend_part into feature/eslint-import-sort

commit aa8b781f036187f45c85dc674460784ee3eab89a
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Sat Mar 23 13:16:09 2024 +0700

    fix: поправил название constant в импорте, добавил правило пустой строки после директив

commit 69a6497d07b4793560cb4b191aa99bbe9a2310da
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Sat Mar 23 01:35:51 2024 +0300

    Поправил имя constants, поправил по мелочи файлы

commit 16d24da268085e51cb5fa7ea8d6c2c54e0843e8b
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Sat Mar 23 01:01:35 2024 +0300

    Поправил алиасы, сделал пробел после use client

commit 359e06a0baf2441390a1627095fdb66f7fd434e4
Merge: 6ba9071 14b64e3
Author: Glushenkov Konstantin <glukosalex@gmail.com>
Date:   Thu Mar 21 22:05:02 2024 +0700

    Merge branch 'develop' into feature/eslint-import-sort

commit 14b64e3888846e4ca67ea3c044f2db2a9003d592
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Thu Mar 21 00:09:24 2024 +0300

    Сделал один файл Constans, удалил в форме картинки по 7 МБ, скорректировал компонент Input, в константах поправил уникальность id

commit 1f59b272123ef858cdc0aab8b3dfaa76caf56b85
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Wed Mar 20 22:56:35 2024 +0300

    Итоговая версия слайдера реализованного на SwiperJS

commit 6ba90717a5982463bd786474dce1c228b066de5d
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 23:07:03 2024 +0700

    fix: fix path to images

commit e9c3a247af466cf70f1b6aaa6d0e920a1cfb2343
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 23:02:57 2024 +0700

    style: fix parent paths to alias '@'

commit d70a6801eefa56f4386efacc6f952c08d997ed0b
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 23:02:04 2024 +0700

    style: add eslint rule import/newline-after-import

commit fba2499ac5f1e5d85bce8af328f4d8ad8e8ca0b7
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 22:33:12 2024 +0700

    style: auto sort all imports

commit c56c179b63378a53ba1fce00b7b4064aa7343af9
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 22:32:31 2024 +0700

    style: config eslint and prettier

commit f969aecdaff18e5ac8ad0b770885a0bfc63579c9
Author: GluKosAlex <glukosalex@gmail.com>
Date:   Wed Mar 20 22:31:08 2024 +0700

    build: install simple-import-sort plugin

commit 90f1447ce26db2849bc5423ef40855269bfeef4b
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Wed Mar 20 01:08:32 2024 +0300

    Убрать лишние компоненты, оставить только тестовый Swiper

commit d754e22db64004869905070aa81191a1555a569e
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Mon Mar 18 23:13:30 2024 +0300

    Настройка брейкпоинтов. Осталось сделать на 480

commit 4faf5d0b58ff9a3fc00103f815a264fd057a8e3e
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Mon Mar 18 00:42:10 2024 +0300

    Настройка брейкпоинтов. Осталось сделать на 394

commit 5b44981b94b962fdd6b02befce070b318083308c
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Wed Mar 13 23:21:07 2024 +0300

    Настройка брейкпоинтов. Осталось сделать на 394, 768 + почистить мусор

commit 80dfed37d9481c41d512e5fded3f628557bee7c6
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Wed Mar 13 00:48:44 2024 +0300

    Настройка брейкпоинтов и стилей на разных разрешениях

commit 7ed8f5479ec688f789322f906bdcee0a1da06c2d
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Tue Mar 12 00:10:07 2024 +0300

    Переделал слайдер отзывов на основе библиотеки Swiper.js

commit 00431c61d3c0c6ade64f35cd7597530ab06b2f18
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Mon Mar 4 23:38:21 2024 +0300

    Создал компонент, настроил функциональность. Сделал стили. На проверку

commit 3cb369c5c15035999b5ddcd223071b9fceab4f52
Author: Pavel Demin <pavlikdemin@yandex.ru>
Date:   Mon Mar 4 00:10:56 2024 +0300

    Создал компонент, настроил функциональность. Нужно еще доделать стили на разных разрешениях экрана

spacelessvoid committed:
обновил брейкпойнт в useCardCount

spacelessvoid committed:
Merge branch 'develop' into feature/pagination
